sound/soc/sof/ops.h

```diff
@@ -387,10 +387,47 @@ static inline int snd_sof_dma_trace_release(struct snd_sof_dev *sdev)
 
 static inline int snd_sof_dma_trace_trigger(struct snd_sof_dev *sdev, int cmd)
 {
-	if (sof_ops(sdev)->trace_trigger)
-		return sof_ops(sdev)->trace_trigger(sdev, cmd);
+	int ret = 0;
 
-	return 0;
+	if (!sdev->dtrace_is_supported)
+		return 0;
+
+	/* FIXME: return 0 or return -EINVAL ? */
+	if (!sof_ops(sdev)->trace_trigger)
+		return 0;
+
+	switch (cmd) {
+	case SNDRV_PCM_TRIGGER_START:
+		/* FIXME: return 0 when dtrace_is_enabled? */
+		if (sdev->dtrace_is_enabled)
+			return 0;
+		if (!sdev->dma_trace_pages)
+			return -EINVAL;
+		ret = sof_ops(sdev)->trace_trigger(sdev, cmd);
+		if (ret < 0) {
+			dev_err(sdev->dev,
+				"error: snd_sof_dma_trace_trigger: start: %d\n", ret);
+		} else {
+			sdev->dtrace_is_enabled = true;
+		}
+		return ret;
+	case SNDRV_PCM_TRIGGER_STOP:
+		if (!sdev->dtrace_is_enabled)
+			return 0;
+		ret = sof_ops(sdev)->trace_trigger(sdev, cmd);
+		if (ret < 0) {
+			dev_err(sdev->dev,
+				"error: snd_sof_dma_trace_trigger: stop: %d\n", ret);
+		} else {
+			sdev->dtrace_is_enabled = false;
+		}
+		return ret;
+	default:
+		ret = sof_ops(sdev)->trace_trigger(sdev, cmd);
```

Comment:
is this correct? what other commands do we expect for trace? we should just break and return 0 right?

Reply:
I call the trace_trigger() to let trace_trigger handle such unexpected cmd because this is expandable. In case we need other cmd for dma trace in the future, this way should be the most friendly.

Reply:
I think this is aligned with the existing implementation.